as/plugins/org.jboss.ide.eclipse.as.wtp.core/src/org/jboss/ide/eclipse/as/core/server/UnitedServerListenerManager.java

```diff
@@ -71,37 +95,61 @@ public static boolean isJBossServer(IServer server) {
 	}
 
 
-	public synchronized void addListener(UnitedServerListener listener) {
-		if( !list.contains(listener)) {
-			list.add(listener);
-			// united listener manager is already initialized, so init this new listener individually
-			if( isInitialized )
-				initializeListener(listener);
+	public void addListener(UnitedServerListener listener) {
+		boolean requiresInit = false;
+		synchronized(this) {
+			if( !list.contains(listener)) {
+				list.add(listener);
+				requiresInit = isInitialized;
+			}
+		}
+		// united listener manager is already initialized, so init this new listener individually
+		if( requiresInit ) {
+			initializeListener(listener);
 		}
 	}
 
-	private synchronized void initializeCurrentListeners() {
-		Iterator<UnitedServerListener> it = list.iterator();
-		while(it.hasNext()) {
-			initializeListener(it.next());
+	private void initializeCurrentListeners() {
+		// Call the init method on all registered UnitedServerListener
+		UnitedServerListener[] all = null;
+		synchronized(this) {
+			all = getListeners();

```

Comment:
if list is an instance of CopyOnWriteArrayList, you don't need to make multiple copies internally, i.e call list directly instead of calling getListeners() every time. this will be more efficient (less array copies)